Refactor templates package

yuri-tceretian opened the pull request:
This pull request refactors package `templates` and makes it more isolate exposing less implementation details outside, making private as many functions as possible.
The goal of this pull request is to prepare the package to provide a single way to obtain Grafana or Prometheus templates, which will be implemented later.

Key changes:
- introduced template `Kind`, which can be either Grafana or Prometheus. 
- refactored fromContent function to build kind-agnostic templates.
- moved functions to process templates from notify package and made them public. 
- updated function TemplateFromContent to accept TemplateDefinition instead of list of template content

yuri-tceretian committed:
introduce template Kind

add default options and templates per kind

yuri-tceretian committed:
refactor: move ParseTestTemplate and TemplateFromContent to templates

yuri-tceretian committed:
make NewTemplate private

yuri-tceretian committed:
remove newTemplate

yuri-tceretian committed:
remove newRawTemplate

yuri-tceretian committed:
make fromContent private

yuri-tceretian committed:
update fromContent to create upstream template by default

update usages to keep creating Grafana templates

yuri-tceretian committed:
make topTemplates private

yuri-tceretian committed:
refactor TemplateFromContent to accept TemplateDefinition

yuri-tceretian committed:
fmt

stevesg commented on notify/receivers.go:
```diff
@@ -109,10 +110,8 @@ func (e IntegrationTimeoutError) Error() string {
-	for _, tc := range am.templates {
-		tmpls = append(tmpls, tc.Template)
-	}
+	tmpls := make([]templates.TemplateDefinition, 0, len(am.templates))
```
```suggestion
	tmpls := make([]templates.TemplateDefinition, len(am.templates))
```

stevesg commented on notify/templates.go:
```diff
@@ -101,59 +99,15 @@ func (am *GrafanaAlertmanager) TestTemplate(ctx context.Context, c TestTemplates
-		seen[tc.Name] = struct{}{}
-	}
-
+	tmpls := make([]templates.TemplateDefinition, 0, len(am.templates))
```
```suggestion
	tmpls := make([]templates.TemplateDefinition, len(am.templates))
```

stevesg reviewed (commented):
First pass LGTM

yuri-tceretian committed:
fix stupid mistake

yuri-tceretian committed:
Merge branch 'main' into yuri-tceretian/template-refactor

yuri-tceretian committed:
Merge branch 'main' into yuri-tceretian/template-refactor

yuri-tceretian committed:
Merge remote-tracking branch 'origin/main' into yuri-tceretian/template-refactor

# Conflicts:
#	templates/default_template_test.go

yuri-tceretian committed:
improve kinds

drop Template from name (because package assumes)
make it int and add some validation

yuri-tceretian committed:
replace static maps with functions

they are more extensible and it will be needed in future

yuri-tceretian committed:
rename PrometheusKind to MimirKind

JacobsonMT commented on notify/grafana_alertmanager.go:
```diff
@@ -469,14 +469,14 @@ func newTestReceiversResult(alert types.Alert, results []result, receivers []*AP
 	testAlert := newTestAlert(c, now, now)
 
-	tmpl, err := templateFromContent(tmpls, externalURL)
+	tmpl, err := templates.TemplateFromTemplateDefinitions(tmpls, log.NewNopLogger(), externalURL)
```
NIT: Probably fine to add a `logger log.Logger` parameter to this function. It's only external use is from Mimir's `func (am *Alertmanager) TestReceiversHandler` which has a logger handy to pass in.

JacobsonMT commented on notify/grafana_alertmanager.go:
```diff
@@ -580,30 +580,35 @@ func TestTemplate(ctx context.Context, c TestTemplatesConfigBodyParams, tmpls []
 		newTextTmpl = text
 	}
-	newTmpl, err := templateFromContent(templateContents, externalURL, captureTemplate)
+	newTmpl, err := templates.TemplateFromTemplateDefinitions(templateContents, log.NewNopLogger(), externalURL, captureTemplate)
```
This function has a logger parameter already, no need to use `log.NewNopLogger()`

JacobsonMT commented on templates/util.go:
```diff
@@ -106,3 +112,53 @@ func checkNode(node parse.Node, executedTmpls map[string]struct{}) {
+// ParseTestTemplate parses the test template and returns the top-level definitions that should be interpolated as results.
+func ParseTestTemplate(name string, text string) ([]string, error) {
+	var tmpl *tmpltext.Template
+	var capture template.Option = func(text *tmpltext.Template, _ *tmplhtml.Template) {
```
NIT: Tempting to get rid of all these awkward `template.Option` hacks used to capture the underlying `tmpltext` at this point by just creating our own `Template` type in `fromContent` that stores a reference to the `tmpltext`. Not necessary though as I imagine the next iteration implementing the factory might change this bit as well.

yuri-tceretian replied:
I will see what can be done in the follow-up PRs.

JacobsonMT reviewed (approved):
LGTM, great refactor! :rocket: All comments are non-blocking.

yuri-tceretian committed:
use proper loggers

yuri-tceretian committed:
Merge branch 'main' into yuri-tceretian/template-refactor
